backend/mem/file.go

```diff
@@ -0,0 +1,497 @@
+	if werr != nil {
+		return nil, werr
+	}
+	cerr := f.CopyToFile(newFile) //copying over the data
+	if cerr != nil {
+		return nil, cerr
+	}
+
+	derr := f.Delete()
+	if derr != nil {
+		return nil, derr
+	}
+	return newFile, nil
+}
+
+/*
+MoveToFile creates a newFile, and moves it to "file".
+If names are same, "file" is deleted and newFile takes its place.
+The receiver is always deleted (since it's being "moved")
+*/
+
+func (f *File) MoveToFile(file vfs.File) error {
+
+	if f.Name() == file.Name() {
+
+		loc := file.Location()
+
+		derr := file.Delete()
+		if derr != nil {
+			return derr
+		}
+
+		newFile, nerr := loc.NewFile(f.Name())
+		if nerr != nil {
+			return nerr
+		}
+
+		_, werr := newFile.Write(make([]byte, 0))
+		if werr != nil {
+			return werr
+		}
+		copyErr := f.CopyToFile(newFile)
+		if copyErr != nil {
+			return copyErr
+		}
+
+		return f.Delete()
+	}
+
+	newFile, nerr2 := file.Location().NewFile(f.Name())
+	if nerr2 != nil {
+		return nerr2
+	}
+	_, werr := newFile.Write(make([]byte, 0))
+	if werr != nil {
+		return werr
+	}
+	copyErr := f.CopyToFile(newFile)
+	if copyErr != nil {
+		return copyErr
+	}
+
+	return f.Delete()
+
+}
+
+/*
+Delete removes the file from the fs. Sets it path in the fsMap to nil,
+ and also nils the file's members
+*/
+func (f *File) Delete() error {
+
+	if ex, _ := f.Exists(); !ex {
+		return doesNotExist()
+	}
+
+	if f == nil {
+		return errors.New("Cannot operate on a nil file reference")
+	}
+
+	loc := f.Location().(*Location)
+	mapRef := &loc.fileSystem.fsMap
+	if _, ok := (*mapRef)[loc.Volume()]; ok { //if there are keys at this volume
+		if thisObj, ok2 := (*mapRef)[loc.Volume()][f.Path()]; ok2 { //checking for the object that should contain the file at this key
+			str := f.Path()
+			file := thisObj.i.(*File) //casting a file to the object's "i" interface
+			file.exists = false
+			file = nil
+			thisObj.i = nil
+			thisObj = nil
+			(*mapRef)[loc.Volume()][str] = nil //setting that key to nil so it truly no longer lives on this system
+		}
+	}
+
+	return nil
+}
+
+//newFile creates an in-mem vfs file given the name then returns it
+func newFile(name string) (*File, error) {
+	return &File{
+		lastModified: time.Now(), name: name, cursor: 0,
+		isOpen: false, exists: false,
+	}, nil
+
+}
+
+//LastModified simply returns the file's lastModified, if the file exists
+func (f *File) LastModified() (*time.Time, error) {
+
+	existence, err := f.Exists()
+
+	if existence {
+		return &f.lastModified, err
+	}
+	return nil, err
+}
+
+//Size returns the size of the file contents.  In our case, the length of the file's byte slice
+func (f *File) Size() (uint64, error) {
+
+	return uint64(len(f.fileContents)), nil
+
+}
+
+//Path returns the absolute path to the file
+func (f *File) Path() string {
+
+	return path.Join(f.Location().Path(), f.name)
+
+}
+
+//Name returns the file's base name
+func (f *File) Name() string {
+	//if file exists
```

Comment:
what does this comment refer to? it is returning f.name regardless of file existence. Remove, if it no longer applies.